src/lambda_function/raw_sync/app.py

```diff
@@ -0,0 +1,518 @@
+located near the end of a zip archive, verification can be done extremely quickly
+and without needing to download most of the export.
+
+If a JSON file from an export is found to not have a corresponding object in the raw bucket,
+the export is submitted to the raw Lambda (via the dispatch SNS topic) for processing.
+"""
+
+import json
+import logging
+import os
+import struct
+import zipfile
+from collections import defaultdict
+from io import BytesIO
+from typing import Optional
+
+import boto3
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+
+def lambda_handler(event: dict, context: dict) -> None:
+    """
+    Entrypoint for this Lambda.
+
+    Args:
+        event (dict)
+        context (dict): Information about the runtime environment and
+            the current invocation
+    """
+    s3_client = boto3.client("s3")
+    input_bucket = os.environ.get("INPUT_S3_BUCKET")
+    input_key_prefix = os.environ.get("INPUT_S3_KEY_PREFIX")
+    raw_bucket = os.environ.get("RAW_S3_BUCKET")
+    raw_key_prefix = os.environ.get("RAW_S3_KEY_PREFIX")
+    dispatch_sns_arn = os.environ.get("SNS_TOPIC_ARN")
+    main(
+        event=event,
+        s3_client=s3_client,
+        input_bucket=input_bucket,
+        input_key_prefix=input_key_prefix,
+        raw_bucket=raw_bucket,
+        raw_key_prefix=raw_key_prefix,
+        dispatch_sns_arn=dispatch_sns_arn,
+    )
+
+
+def append_s3_key(key: str, key_format: str, result: dict) -> None:
+    """
+    Organizes an S3 object key by appending it to the appropriate entry in the result dictionary
+
+    This is a helper function for `list_s3_objects`.
+
+    Args:
+        key (str): The S3 object key to process.
+        key_format (str): The format of the key, either "raw" or "input".
+        result (dict): The dictionary where keys are appended. For the "raw" format, it is a
+                       nested dictionary structured as result[data_type][cohort]. For "input",
+                       it is structured as result[cohort].
+
+    Returns:
+        None
+    """
+    if not key.endswith("/"):  # Ignore keys that represent "folders"
+        key_components = key.split("/")
+        if key_format == "raw":
+            try:
+                data_type = next(
+                    part.split("=")[1]
+                    for part in key_components
+                    if part.startswith("dataset=")
+                )
+                cohort = next(
+                    part.split("=")[1]
+                    for part in key_components
+                    if part.startswith("cohort=")
+                )
+                result[data_type][cohort].append(key)
+            except StopIteration:
+                # Skip keys that don't match the expected pattern
+                return
+        elif key_format == "input" and len(key_components) == 3:
+            cohort = key_components[1]
+            result[cohort].append(key)
```

Comment:
Does this function not return anything?  Are you taking advantage of the mutable dict that you pass in through memory?  I feel like that adds vulnerability in the code

Reply:
The function modifies the `result` dict in-memory.

> I feel like that adds vulnerability in the code

How so? I could make a copy of the dict, update the copy, and then return -- but what's the benefit?

Reply:
Actually, I did think of a small benefit. Returning a copy makes the variable update in `list_s3_objects` more explicit, and creating a shallow copy within the function can be done without any additional imports and with just one extra line of code.

Reply:
This is probably more relevant when making mutable default arguments, but I prefer the way you've written it to make the function self contained
https://docs.python-guide.org/writing/gotchas/#:~:text=Python's%20default%20arguments%20are%20evaluated,to%20the%20function%20as%20well.